src/vma/dev/qp_mgr_eth_mlx5.cpp

```diff
@@ -75,6 +75,80 @@ static inline uint64_t align_to_WQEBB_up(uint64_t val)
 	return ((val+4-1)>>2)<<2;
 }
 
+
+//! Copying src to BlueFlame register buffer by Write Combining cnt WQEBBs
+// Avoid using memcpy() to copy to BlueFlame page, since memcpy()
+// implementations may use move-string-buffer assembler instructions,
+// which do not guarantee order of copying.
+static void copy_bf_nt(uint64_t *bf_dst, uint64_t *src, int cnt)
+{
+	while (cnt--) {
+		COPY_64B_NT(bf_dst, src);
+	}
+}
+
+static void copy_bf_asm(uint64_t *bf_dst, uint64_t *src, int cnt)
+{
+	while (cnt--) {
+		COPY_64B_ASM(bf_dst, src);
+	}
+}
+
+//! Copying two data chunks in wrap around case to BlueFlame register buffer
+//  by Write Combining
+static void copy_bf2_nt(uint64_t* bf_dst, uint64_t* src_bottom, uint64_t* src_top, int cnt_bottom, int cnt_top)
+{
+	while (cnt_bottom--) {
+		COPY_64B_NT(bf_dst, src_bottom);
+	}
+	while (cnt_top--) {
+		COPY_64B_NT(bf_dst, src_top);
+	}
+}
+
+static void copy_bf2_asm(uint64_t* bf_dst, uint64_t* src_bottom, uint64_t* src_top, int cnt_bottom, int cnt_top)
+{
+	while (cnt_bottom--) {
+		COPY_64B_ASM(bf_dst, src_bottom);
+	}
+	while (cnt_top--) {
+		COPY_64B_ASM(bf_dst, src_top);
+	}
+}
+
+qp_mgr_eth_mlx5::qp_mgr_eth_mlx5(const ring_simple* p_ring,
+		const ib_ctx_handler* p_context, const uint8_t port_num,
+		struct ibv_comp_channel* p_rx_comp_event_channel,
+		const uint32_t tx_num_wr, const uint16_t vlan, bool call_configure):
+	qp_mgr_eth(p_ring, p_context, port_num, p_rx_comp_event_channel, tx_num_wr, vlan, false)
+	,m_hw_qp(NULL)
+	,m_sq_wqe_idx_to_wrid(NULL)
+	,m_sq_wqes(NULL)
+	,m_sq_wqe_hot(NULL)
+	,m_sq_wqes_end(NULL)
+	,m_sq_db(NULL)
+	,m_sq_bf_reg(NULL)
+	,m_qp_num(0)
+	,m_sq_wqe_hot_index(0)
+	,m_sq_bf_offset(0)
+	,m_sq_bf_buf_size(0)
+	,m_sq_wqe_counter(0)
+	,m_dm_enabled(0)
+{
+	if (call_configure && configure(p_rx_comp_event_channel)) {
+		throw_vma_exception("failed creating qp_mgr_eth");
+	}
+
+	copy_bf = &copy_bf_asm;
+	copy_bf2 = &copy_bf2_asm;
+	if (mce_sys_var::HYPER_KVM == safe_mce_sys().hypervisor) {
```

Comment:
This verification is to "general" - we must understand how mlx5 detect this limitation...